macosx/kext/AudioEngine.cpp

```diff
@@ -460,13 +461,67 @@ IOAudioStream *kXAudioEngine::createNewAudioStream(int chn,IOAudioStreamDirectio
 			if(sampleBuffer && sampleBufferSize)
 				audioStream->setSampleBuffer(sampleBuffer, sampleBufferSize);
             
-				// This device only allows a single format and a choice of 2 different sample rates
-            rate.fraction = 0;
-            rate.whole = sampling_rate;
-            audioStream->addAvailableFormat(&format, &rate, &rate);
-            
-				// Finally, the IOAudioStream's current format needs to be indicated
-            audioStream->setFormat(&format);
+            if(hw->is_10k2){
+                rate.fraction = 0;
+                rate.whole = sampling_rate;
+                format.fBitDepth = 24;
+                format.fBitWidth = 32;
+                audioStream->addAvailableFormat(&format, &rate, &rate);
+                
+                // Finally, the IOAudioStream's current format needs to be indicated
+                audioStream->setFormat(&format);
+                
+                rate.whole = 44100;
+                format.fBitDepth = 24;
+                format.fBitWidth = 32;
+                audioStream->addAvailableFormat(&format, &rate, &rate);
+                
+                if(!hw->is_edsp){
+                    rate.whole = 88200;
+                    format.fBitDepth = 24;
+                    format.fBitWidth = 32;
+                    audioStream->addAvailableFormat(&format, &rate, &rate);
+                    
+                    rate.whole = 96000;
+                    format.fBitDepth = 24;
+                    format.fBitWidth = 32;
+                    audioStream->addAvailableFormat(&format, &rate, &rate);
+                    
+                    rate.whole = 176400;
+                    format.fBitDepth = 24;
+                    format.fBitWidth = 32;
+                    audioStream->addAvailableFormat(&format, &rate, &rate);
+                }
+            }else{
+                rate.fraction = 0;
+                rate.whole = sampling_rate;
+                format.fBitDepth = 16;
+                format.fBitWidth = 16;
+                audioStream->addAvailableFormat(&format, &rate, &rate);
+                
+                // Finally, the IOAudioStream's current format needs to be indicated
+                audioStream->setFormat(&format);
+                
+                rate.whole = 44100;
```

Comment:
44.1 is not supported on 10k1 devices without hardware SRC
hardware SRC is much worse than modern software SRC implemented by apple in CoreAudio
I doubt we need to support 44.1 on 10k1 devices
The same applies to 88.2, 96 and 176.4 below

Reply:
I do not understand how this algorithm works (voicetable[].param.initial_pitch/pitch_target)

Reply:
There are two different formulas used for calculating pitch and sampling rate. Pitch is relative to 48000 and is measured in 'cents'.